add ability for Harvester to artificially wait after unloading

Shrooblord opened the pull request:
With this addition, `Harvester`s can be told to explicitly wait (at the Refinery) for a number of ticks before continuing to the next Activity (e.g. returning to a Tiberium field).

E.g.
```yaml
Harvester:
	DeliveryActors: proc
	Capacity: 50
	Resources: Tiberium, BlueTiberium
	...
	UnloadFinishedDelay: 30
	...
	EmptyCondition: no-tiberium
```

Context:
For some more complex behaviour for a new unit deriving from `Harvester`, I required this delay. However, instead of cloning the entire stack of `Harvester` classes and related functions, Interfaces, etc., I added these couple of lines to the base code. I figured maybe someone else out there would like a configurable delay for their (new) `Harvester`-like units.

Housekeeping:
- [x] Read guidelines for contributing
- [x] `make test`
- [x] `make check`

Shrooblord committed:
add ability for Harvester to artificially wait after unloading

PunkPun commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -266,11 +272,16 @@ public virtual bool TickUnload(Actor self, Actor proc)
+			// Wait until artificial delay time has passed
+			if (--currentUnloadFinishedDelayTicks > 0)
+				return false;
+
```
This isn't the right way to implement it. You should be using activities, see from where the TickUnload function is called. 

One way would be calling `QueueChild(new Wait(UnloadFinishedDelay));` when the loop ends in `HarvesterDockSequence`

PunkPun replied:
Though actually that might be wrong as well. Your implementation of `currentUnloadFinishedDelayTicks` handling could just be moved to  `HarvesterDockSequence` and I think it would be fine

Shrooblord replied:
Oh, alright. I was copying the functionality already present in-code there. I'll look into moving it into a more conventional place, thanks!

Shrooblord committed:
move functionality into HarvesterDockSequence.cs

Shrooblord commented:
@PunkPun how's that? I tested it in-game by first testing vanilla, and then adding the `UnloadFinishedDelay: 120` key to the `rules/shared-vehicles.yaml/HARV` definition, and it's working as intended on my end.

PunkPun commented:
I'm just wondering, where do we want the delay? Before `Drag` activity or after, and do we want to consider the harvester as still docked or undocked while waiting? If waiting after everything is done, queueing a Wait activity would be best, if somewhere inbetween, I'd like to hear your usecase

Shrooblord commented:
For me personally, I introduced the change in my mod because I have a `Harvester` that also has the ability to Deploy into a `Building`, but only when it has an empty Tiberium storage. In regular `Harvester` behaviour, the time you had as a player to respond to the unit being done offloading its Tib (which in this case was 'to itself' because it's also its own `Refinery`) and it immediately going off on another resource run, felt a bit like you had to spam-click to get the unit to do what you wanted. That doesn't feel very good, so I added this delay to give the player more time to react to the "empty storage -> could Deploy now if you want" condition.

But it could also be used for purely aesthetical reasons; being able to more finely control the timing of various stages of loading, docking, and undocking, are always nice-to-haves as a mod author. That, however, is super niche and I recognise that.

Where it 'makes most sense' to live, I leave up to you. I'm not familiar enough with the codebase yet to know what feels most 'OpenRA-ey' if that makes sense? x)

PunkPun commented:
It might be worth scope creeping this into adding a configurable before unloading delay. Currently it's hardcoded to 10 ticks https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/DeliverResources.cs#L74

Shrooblord commented:
Yes! Stuff like that are definitely nice-to-haves. I'd be fine implementing those myself if my current approach is approved -- then I'll just copy the same behaviour to the other sequences of docking using similar code.

Shrooblord commented:
Ah, I did just notice that moving the code has changed the delay behaviour from "giving the player a window to react" to "obligate delay" i.e. now the unit doesn't instantly respond to the Deploy command anymore, probably because it's still got the Undock Action that it's locked into. That's not a deal-breaker for me, but it is a sign that the code doesn't behave 100% the same as before. YMMV

PunkPun commented:
Ah, in that case the delay should be after fully finishing the `HarvesterDockSequence` activity. Call 
```c#
QueueChild(new Wait(Harv.Info.UnloadFinishedDelay));
```

In https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/HarvesterDockSequence.cs#L94 just before `return true;` Perhaps could also add a check so we would only queue it if value > 0

`UnloadFinishedDelay` could be reworded to something like `WaitAfterUnload`
